src/apdu/dispatcher.c

```diff
@@ -72,10 +76,50 @@ int apdu_dispatcher(const command_t *cmd) {
                 // The data length is variable on each stage which requires data.
                 // so we will just test for existence or not of data
                 return io_send_sw(SW_WRONG_DATA_LENGTH);
-            } else if (cmd->p1 == SIGN_TX_P1_DONE && cmd->data) {
             }
 
             return handler_sign_tx(&buf, (sing_tx_stage_e) cmd->p1, cmd->p2);
+        case SIGN_TOKEN_DATA:
+            if (cmd->p1 != 0 || cmd->p2 != 0) {
+                return io_send_sw(SW_WRONG_P1P2);
+            }
+
+            buf.ptr = cmd->data;
+            buf.size = cmd->lc;
+            buf.offset = 0;
+
+            return handler_sign_token_data(&buf);
+        case SEND_TOKEN_DATA:
+            if (cmd->p2 != 0) {
```

Comment:
The p1 and p2 are arbitrary and is up to the handler to interpret the meaning. On this command the p2 is unnecessary, so we only accept 0 (could be anything really) and p1 is the index of the token being sent, this was done to clean the token registry when the wallet is sending the first token of the transaction (p1 equals 0).

The value of p1 (other than 0) has no consequence, the first idea was to use it as the index on the registry but this was better done internally by having the length of saved tokens on the global context.
This is slightly better because we can check the token_data of an output and if it references an index larger than what we have saved we can reject the tx as the context was not initialized correctly.

Reply:
Got it!